### PR TITLE
feat: add install type to context

### PR DESCRIPTION
**Pull request by saikumarrs**
## PR Description

We're now sending a new context field `installType` in `context.app` object which indicates whether the installation type is CDN or NPM.
Based on this, the fixtures in sanity suite have been modified accordingly.

Other minor updates:
- Slack notifications for NPM publishes are sent only when a package is actually published.
- Hyperlink text in Slack notifications tailed for different environment deployments.
- Fixed the installation type value sent to Bugsnag.
- Fixed the app version sent to Bugsnag.
- For user ID in Bugsnag payload, we prefer source ID over write key.
- Fixed an issue in the URL validation to cover the scenario where the 'URL' object is not available in the browsers (legacy).

## Linear task (optional)

https://linear.app/rudderstack/issue/SDK-1477/send-the-installation-type-in-the-event-context-js-sdk

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [x] Chrome
- [ ] Firefox
- [x] IE11

## Sanity Suite

- [x] All sanity suite test cases pass locally

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Added `installType` field to the `AppInfo` type for better context in analytics.

- **Bug Fixes**
  - Enhanced error handling and context information in Bugsnag utilities.

- **Chores**
  - Updated deployment workflows to include version information based on the environment.
  - Improved NPM version retrieval and Slack notifications in deployment workflows.

- **Tests**
  - Added new test cases for URL validation and Bugsnag utilities, including setting and deleting global variables.

### CODE REVIEW

File: packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts [hunk 1]
```diff
@@ -166,11 +174,11 @@ describe('Bugsnag utilities', () => {
         errorMessage: 'test error message',
       };
 
-      enhanceErrorEventMutator(event, 'dummyMetadataVal');
+      enhanceErrorEventMutator(event);
 
       expect(event.metadata).toEqual({
         source: {
-          metadataSource: 'dummyMetadataVal',
+          snippetVersion: '3.0.0',
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts:181]: Invalid suggestion.

File: packages/analytics-js-plugins/__tests__/bugsnag/index.test.ts [hunk 2]
```diff
@@ -14,6 +14,14 @@ describe('Plugin - Bugsnag', () => {
...
+      app: signal({
+        name: 'test-app',
+        namespace: 'test-namespace',
+        version: '1.0.0',
+        installType: 'npm',
+      }),
+    },
   };
 
   let state: any;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/__tests__/bugsnag/index.test.ts:27]: It's fine for this.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts [hunk 3]
```diff
@@ -60,9 +60,8 @@ const isRudderSDKError = (event: any) => {
   );
 };
 
-const enhanceErrorEventMutator = (event: any, metadataSource: any) => {
+const enhanceErrorEventMutator = (event: any) => {
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts]: Done.

File: packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts [hunk 4]
```diff
@@ -19,6 +19,14 @@ import {
 } from '../../src/bugsnag/utils';
 import { server } from '../../__fixtures__/msw.server';
 
+beforeEach(() => {
+  window.RudderSnippetVersion = '3.0.0';
+});
+
+afterEach(() => {
+  delete window.RudderSnippetVersion;
+});
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts]: Done.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts [hunk 5]
```diff
@@ -60,9 +60,8 @@ const isRudderSDKError = (event: any) => {
   );
 };
 
-const enhanceErrorEventMutator = (event: any, metadataSource: any) => {
+const enhanceErrorEventMutator = (event: BugsnagLib.Report) => {
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts]: Done.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts [hunk 6]
```diff
@@ -60,9 +60,8 @@ const isRudderSDKError = (event: any) => {
   );
 };
 
-const enhanceErrorEventMutator = (event: any, metadataSource: any) => {
+const enhanceErrorEventMutator = (event: BugsnagLib.Report): void => {
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts:63]: Invalid suggestion.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts [hunk 7]
```diff
@@ -114,22 +111,22 @@ const getNewClient = (state: ApplicationState, logger?: ILogger): BugsnagLib.Cli
 
   const clientConfig: BugsnagLib.IConfig = {
     apiKey: API_KEY,
-    appVersion: '__PACKAGE_VERSION__', // Set SDK version as the app version from build config
+    appVersion: state.context.app.value.version,
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts:114]: Invalid suggestion.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts [hunk 8]
```diff
@@ -81,17 +80,16 @@ const enhanceErrorEventMutator = (event: any, metadataSource: any) => {
...
+  (): BugsnagLib.BeforeSend =>
+  (event: BugsnagLib.Report): boolean => {
     try {
       // Discard the event if it's not originated at the SDK
       if (!isRudderSDKError(event)) {
         return false;
       }
 
-      enhanceErrorEventMutator(event, metadataSource);
+      enhanceErrorEventMutator(event);
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts:92]: We can ignore those errors.

File: packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts [hunk 9]
```diff
@@ -166,11 +176,11 @@ describe('Bugsnag utilities', () => {
         errorMessage: 'test error message',
       };
 
-      enhanceErrorEventMutator(event, 'dummyMetadataVal');
+      enhanceErrorEventMutator(event);
 
       expect(event.metadata).toEqual({
         source: {
-          metadataSource: 'dummyMetadataVal',
+          snippetVersion: '3.0.0',
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts:183]: Invalid suggestion.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts [hunk 10]
```diff
@@ -167,11 +164,11 @@ const loadBugsnagSDK = (externalSrcLoader: IExternalSrcLoader, logger?: ILogger)
 
 const initBugsnagClient = (
   state: ApplicationState,
-  promiseResolve: (value: any) => void,
+  promiseResolve: (value: BugsnagLib.Client) => void,
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts:167]: Done.

File: packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts [hunk 11]
```diff
@@ -206,14 +216,26 @@ describe('Bugsnag utilities', () => {
...
   describe('initBugsnagClient', () => {
-    const state = {
-      source: signal({
-        id: 'dummy-source-id',
-      }),
-      lifecycle: {
-        writeKey: signal('dummy-write-key'),
-      },
-    };
+    let state;
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts]: Done.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts]: Done.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts [hunk 12]
```diff
@@ -114,22 +116,22 @@ const getNewClient = (state: ApplicationState, logger?: ILogger): BugsnagLib.Cli
 
   const clientConfig: BugsnagLib.IConfig = {
     apiKey: API_KEY,
-    appVersion: '__PACKAGE_VERSION__', // Set SDK version as the app version from build config
+    appVersion: state.context.app.value.version,
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts:119]: Invalid suggestion.

File: packages/analytics-js-common/src/utilities/url.ts [hunk 13]
```diff
@@ -9,15 +9,19 @@ const removeDuplicateSlashes = (str: string): string => str.replace(/\/{2,}/g, '
  * @returns true if `url` is valid and false otherwise
  */
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-common/src/utilities/url.ts:10]: Invalid suggestion.

File: packages/analytics-js-plugins/src/bugsnag/utils.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/src/bugsnag/utils.ts:119]: The type is already defined as ApplicationState.

File: packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts [hunk 14]
```diff
@@ -166,11 +196,11 @@ describe('Bugsnag utilities', () => {
         errorMessage: 'test error message',
       };
 
-      enhanceErrorEventMutator(event, 'dummyMetadataVal');
+      enhanceErrorEventMutator(event);
 
       expect(event.metadata).toEqual({
         source: {
-          metadataSource: 'dummyMetadataVal',
+          snippetVersion: '3.0.0',
```
[the remark this answers is not in the document]
Comment (saikumarrs) [reply packages/analytics-js-plugins/__tests__/bugsnag/utils.test.ts:203]: Invalid suggestion. The snippet version is not set twice.

**Review by MoumitaM — APPROVED** [review]

**Pull request merged**